File: src/DAO/ScoreCalculator.js
import {predictions} from "./Predictions";

export const calculateScores = () => {
    const scoredPredictions = predictions.map((station) => {

        const maxValues = station.Predictions.reduce((maxValues, prediction) => {
            const newMaxValues = {...maxValues}
            Object.keys(prediction).forEach((key) => {
                const max = maxValues[key] || Number.MIN_SAFE_INTEGER
                const keyPrediction = prediction[key]
                if (prediction[key]) {
                    newMaxValues[key] = Math.max(max, keyPrediction)
                } else {
                    newMaxValues[key] = max
                }

            })
            return newMaxValues
        }, station.Predictions[0])

        const minValues = station.Predictions.reduce((minValues, prediction) => {
            const newMinValues = {}
            Object.keys(prediction).forEach((key) => {

                const min = minValues[key] || Number.MAX_SAFE_INTEGER
                const keyPrediction = prediction[key]
                if (prediction[key]) {
                    newMinValues[key] = Math.min(min, keyPrediction)
                } else {
                    newMinValues[key] = min
                }

            })
            return newMinValues
        }, station.Predictions[0])


        let editedPredictions = station.Predictions.map((prediction) => {
            const cloudScore = prediction.clouds && 10 * (1 - ((prediction.clouds - minValues.clouds) / (maxValues.clouds - minValues.clouds))) || 0
            const windScore = prediction.windSpeed && 10 * (1 - ((prediction.windSpeed - minValues.windSpeed) / (maxValues.windSpeed - minValues.windSpeed))) || 0
            const pressureScore = prediction.airPressure && 5 * ((prediction.airPressure - minValues.airPressure) / (maxValues.airPressure - minValues.airPressure)) || 0
            const humidityScore = prediction.relativeHumidity && 5 * (1 - ((prediction.relativeHumidity - minValues.relativeHumidity) / (maxValues.relativeHumidity - minValues.relativeHumidity))) || 0
            const maxTempScore = prediction.maxAirTemperature && 60 * ((prediction.maxAirTemperature - minValues.maxAirTemperature) / (maxValues.maxAirTemperature - minValues.maxAirTemperature)) || 0
            const minTempScore = prediction.minAirTemperature && 10 * ((prediction.minAirTemperature - minValues.minAirTemperature) / (maxValues.minAirTemperature - minValues.minAirTemperature)) || 0
            return {...prediction, score: cloudScore + windScore + pressureScore + humidityScore + maxTempScore + minTempScore}
        })

        const minScore = editedPredictions.reduce((minScore, prediction) => Math.min(minScore, prediction.score), Number.MAX_SAFE_INTEGER)
        const maxScore = editedPredictions.reduce((maxScore, prediction) => Math.max(maxScore, prediction.score), Number.MIN_SAFE_INTEGER)

        editedPredictions = editedPredictions.map((prediction) => ({...prediction, score: (prediction.score-minScore)/(maxScore-minScore)}))

        return {...station, Predictions: editedPredictions}
    })

    download("scoredPredictions.json", JSON.stringify(scoredPredictions))
}

function download(filename, text) {
    var element = document.createElement('a');
    element.setAttribute('href', 'data:text/plain;charset=utf-8,' + encodeURIComponent(text));
    element.setAttribute('download', filename);

    element.style.display = 'none';
    document.body.appendChild(element);

    element.click();

    document.body.removeChild(element);
}

File: src/DAO/ScoreCalculator.py
# -*- coding: utf-8 -*-


import json
import sys
from functools import reduce
predictions = json.loads(open(r"C:\Users\Juuso Lassila\Documents\pred_results_2019.json").read())

def scoreStation(station):
    
    def getBiggerValues(prediction1, prediction2):
        newMaxValues = prediction1.copy()
        for key in prediction2:
            maxVal = prediction1[key] or sys.float_info.min
            pred2Val = prediction2[key]
            if pred2Val:
                newMaxValues[key] = max(maxVal, pred2Val)
            else:
                newMaxValues[key] = maxVal
        return newMaxValues
    
    maxValues = reduce(getBiggerValues, station["Predictions"], station["Predictions"][0])
    
    def getSmallerValues(prediction1, prediction2):
        newMinValues = prediction1.copy()
        for key in prediction2:
            minVal = prediction1[key] or sys.float_info.max
            pred2Val = prediction2[key]
            if pred2Val:
                newMinValues[key] = min(minVal, pred2Val)
            else:
                newMinValues[key] = minVal
        return newMinValues
    
    minValues = reduce(getSmallerValues, station["Predictions"], station["Predictions"][0])
    
    def normalize(value, maxValue, minValue):
        return (value-minValue)/(maxValue-minValue)
        
    def addScoreToPrediction(prediction):
        maxScore = 0
        def calculateScore(property, weight, inverted=False):
            nonlocal prediction
            nonlocal maxScore
            nonlocal maxValues
            nonlocal minValues
            if property in prediction and prediction[property]:
                maxScore += weight
                if inverted:
                    return (weight * (1-normalize(prediction[property], maxValues[property], minValues[property])))
                else:
                    return (weight * normalize(prediction[property], maxValues[property], minValues[property]))
            return 0
        
        cloudScore = calculateScore("clouds", 10, True)
        windScore = calculateScore("windSpeed", 10, True)
        pressureScore = calculateScore("airPressure", 5)
        humidityScore = calculateScore("relativeHumidity", 5, True)
        maxTempScore = calculateScore("maxAirTemperature", 60)
        minTempScore = calculateScore("minAirTemperature", 10)
        newPrediction = prediction.copy()
        if maxScore != 0:
            newPrediction["score"] = (cloudScore+windScore+pressureScore+humidityScore+maxTempScore+minTempScore)/maxScore
        else: 
            newPrediction["score"] = 0.5
        return newPrediction
    
    editedPredictions = list(map(addScoreToPrediction, station["Predictions"]))
    
    minScore = reduce(lambda m, p: min(m, p), map(lambda p: p["score"], editedPredictions), sys.float_info.max)
    maxScore = reduce(lambda m, p: max(m, p), map(lambda p: p["score"], editedPredictions), sys.float_info.min)
    
    def normalizeScore(prediction):
        newPrediction = prediction.copy()
        newPrediction["score"] = normalize(prediction["score"], maxScore, minScore)
        return newPrediction
    
    editedPredictions = list(map(normalizeScore, editedPredictions))
    
    newStation = station.copy()
    newStation["Predictions"] = editedPredictions
    return newStation
        
        
    
    
predictions = list(map(scoreStation, predictions))
    
with open(r"C:\Users\Juuso Lassila\Documents\scored_pred_results_2019.json", "w") as f:
        json.dump(predictions, f)